### PR TITLE
Piecewise expr must be Basic

### LINKED ISSUE #13223: ITE.to_nnf appears to use Boolean of else arg
I expect both to give the same result in the following:

```
>>> e = ITE(x<1,(1,2),S.NaN)
>>> e.to_nnf().subs(x,2)
True
>>> e.subs(x,2)
NaN
```

The NaN evaluates in a different way than the (1,2) does in an Or: Or(x<1,foo) is True if foo is a Number.

### LINKED ISSUE #13347: incompatible use of binary symbols is allowed in Relationals
While `And(Eq(x, True), x)`  is ok (and should reduce to `x`, writing `And(x < 1, x)` should be disallowed if the args of And should be Booleans since x, being a Boolean in this context, cannot be used in an expression with a real value: neither `x < 1` nor `x + 1`.

Care needs to be taken, however, to not disallow `Eq(x, 1)` since without a context that dictates that x is binary, this will permit any value for x. In the context `And(Eq(x, 1), x)`, however, we can conclude that the result is False since x is a Boolean and `Eq(True, 1)` is False -- though this doesn't evaluate as False and I'm not sure why.

### LINKED ISSUE #13349: solve(Eq(x, True)) raises error
```
>>> solve(Eq(x, True))
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/base/data/home/apps/s~sympy-live-hrd/49.400254913747479351/sympy/sympy/solvers/solvers.py", line 828, in solve
    f[i] = Add(fi.lhs, -fi.rhs, evaluate=False)
TypeError: bad operand type for unary -: 'BooleanTrue'

```

### LINKED ISSUE #13370: Define Basic.__lt__ etc. to raise TypeError on incompatible types
As pointed out by @smichr in Gitter, in Python 2, we get

```py
>>> S.true < 3
False
```

That's because in Python 2, the default comparisons return some arbitrary (but consistent) ordering. In Python 3, we get

```py
>>> S.true < 3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unorderable types: BooleanTrue() < int()
```

We effectively already have consistency between Python 2 and 3 in Expr. A comparison that isn't known to be true or false remains unevaluated, but if you try to use it in a boolean context (like an `if` statement), it raises TypeError. 

I think it would be best to define the comparison operators on Basic to also be consistent, basically to raise TypeError for incomparable types.

Unfortunately, in general, we can't know if a type is incomparable, because it could define a custom comparison operator. It's also not possible to tell if Python is calling `__lt__` from a `<` operator or from a flipped `>` operator. But at least for built in types, like int, float, etc., we can know that they can't compare with non-Expr SymPy types, so we can check that.

### PR DESCRIPTION
**Pull request by smichr**
## Summary of changes

* changes to ExprCondPair
  * expr must be Basic
  * cond must be Boolean or must be able to be reduced to Boolean
  * a cond containing Piecewise will be folded
  * a cond that is a Piecewise (perhaps after folding) will be rewritten as ITE with canonical Relationals
  * a TypeError is raised if the cond is not a Boolean
 * is is used instead of == since ExprCondPair args are Boolean now
* `piecewise_fold` will convert ITE conditions to nnf and simplified if possible
* ITE/And/Or take only Boolean args e.g. `And(x, y)` is okay, `And(2, x + 1, x**2)` is not.
* ITE now pretty prints like Piecewise
* BooleanFunction has a binary_symbols property which returns a set of symbols which can have a boolean/binary value
* as_Basic utility function to return the Basic version of an expression or raise an error if there is none.
* Piecewise._eval_subs is updated to reflect that Piecewise does not contain bool
* `piecewise_fold` now folds the conditions and if the condition (before or after folding) contains Boolean expressions, then it is converted to an unsimplified BooleanFunction.
```
>>> Piecewise((1,Piecewise((x>1,x<3),(x>2,True))))
Piecewise((1, ITE(x < 3, x > 1, x > 2)))
>>> piecewise_fold(_)
Piecewise((1, ((x >= 3) | (x > 1)) & ((x > 2) | (x < 3))))
>>> simplify(_)
Piecewise((1, x > 1))
```
* `as_Boolean` is added to boolalg which behaves like as_Basic
* Boolean gets default `as_set` and `to_nnf` methods
* args of ITE, And and Or must be Boolean (for ITE this required adding a `__new__` definition
* simplify_logic now does simplification of variables before checking if the expression is a BooleanFunction since this may change the input function's type
* Piecewise.rewrite(ITE)  has been added
* conditionals appearing in a solution of a Piecewise function are simplified
* differentiation of Boolean is simply unevaluated for expressions involving relationals in the symbol for which the derivative is requested unless the symbol is a binary symbol
```
>>> (x<1).diff(x)
Derivative(x < 1, x)
>>> And(x<1, x).diff(x)  # result does not depend on whether x is 0 or 1
0
>>> And(x<=1, x).diff(x)  #in this case it does: x =0 gives False while x=1 gives True
1
```
* `func_name` recognizes "<class 'foo'>" when trying to identify `foo`
* PEP8 edits of periodicity tests
* solve and checksol more carefully handle Eq/Ne expressions which have one arg that is True or False; in this case, the other arg must also be a Boolean
* interval solutions from solve are now disabled in Piecewise solving -- when solving for x one should not have x in the solution

(Much of this is split out of #12587)
closes #13223
closes #13255 on which it was built
fixes #13347
fixes #13349 
fixes #13370 
addressed #13350 but doesn't handle case of non-binary symbols; close if master doesn't handle Relationals like `(x**2 < 1).diff(x)`

- [x] check coverage

**REMOVED**
* careful simplification of BooleanFunction is added for univariate expressions
* simplification of complex univariate Relationals has been implemented through the use of sets provided that the involved expressions do not have periodic behavior.
* Relational._eval_simplify has been modified to use the newly added `simplify_relationships` which will try to simplify boolean functions via sets (with care not to add +/-oo when not necessary)

### CODE REVIEW

**Comment by smichr** [comment]
ping @asmeurer -- just double checking that this is the intention (to have nothing but sympified args in objects).

**Comment by asmeurer** [comment]
I think even Tuple in a Piecewise should be avoided, at least when dealing with expressions (but maybe not an error, I'm not sure). If the expression arg isn't an Expr it doesn't make mathematical sense to combine it with another expression.  `Piecewise(((1, 2), x < 0)) + 1` doesn't make sense, but `(Piecewise((1, x < 0)) + 1, Piecewise((2, x < 0)) + 1)` does.

**Comment by smichr** [comment]
Ok. We should think of a Piecewise EXPRESSION not a Piecewise container or a convenient alternative to a nested ITE.

**Comment by asmeurer** [comment]
Probably it could be (ab)used for other functions, but all the methods and simplification functions will assume that things are expressions. For booleans you definitely want to use ITE, as it knows about booleans and has the appropriate boolean methods (like to_nnf). For containers, probably if you are doing this it means you want to be using sets, which have their own ways of representing conditionals (like ConditionSet).

**Comment by asmeurer** [comment]
(by the way I'm +1 to this)

**Comment by smichr** [comment]
OK, let's not commit this yet. There is an XFAILed test that has a condition that has a Piecewise something like:

```
cond = Eq(Piecewise((cos(x), x<0), (a, True)), 0)
```
This could be made into `Piecewise((Eq(cos(x), 0), x<0),(Eq(a, 0), True))` but then the arg is not an Expr, it's a Boolean. So perhaps this is where ITE could be used to write this as `cond = ITE(x<0, Eq(cos(x), 0), Eq(a, 2)).to_nnf()`:

```
And(Or(Eq(a,2),x<0),Or(Eq(cos(x),0),x>=0))
```

**Comment by asmeurer** [comment]
What's wrong with `Eq(Piecewise((cos(x), x<0), (a, True)), 0)`?

**Comment by smichr** [comment]
> What's wrong with ...

Some of the problems that integration of Piecewise has (cf #11045) is that the code to check the domain represented by a condition doesn't fully anticipate what forms the cond part of the ExprCond pair will have. So you end up with AttributeErrors.  So one idea is to make `piecewise_fold` turn piecewise conditions into a Boolean expressions.

**Comment by asmeurer** [comment]
Oh, I see, you have a Piecewise that has another Piecewise in a condition. Is it possible to denest that somehow?

**Comment by asmeurer** [comment]
So say you have `Piecewise((f(x), Eq(Piecewise((cos(x), x<0), (a, True)), 0)), (g(x), True))`

```
⎧          ⎧cos(x)  for x < 0
⎪f(x)  for ⎨                  = 0
⎨          ⎩  a     otherwise
⎪
⎩g(x)          otherwise
```

Isn't this equivalent to `Piecewise((f(x), And(cos(x), x<0)), (f(x), Eq(a, 0)), (g(x), True))`

```
⎧f(x)  for (cos(x) ∨ a = 0) ∧ (a = 0 ∨ x < 0)
⎨
⎩g(x)                otherwise
```

which it looks like it is already simplifying in the same way ITE.to_nnf would:

```
>>> ITE(x < 0, Eq(cos(x), 0), Eq(a, 0)).to_nnf()
(Eq(a, 0) | (x < 0)) & ((x >= 0) | Eq(cos(x), 0))
```

But I understand where you're coming from now. I agree it makes more sense to put a boolean inside a Piecewise. Maybe we should do that. My only worry is that a lot of Piecewise code is written assuming the `expr` parts are `Expr` objects. 

`ITE` could be used but it would need a lot of cleaning up to make it more user friendly (better printing, support for multiple ITEs without nesting), better simplification in terms of higher level logic functions than just And and Or). I remember discussing this on another issue, but when the expressions themselves are booleans, you can simplify them with the conditions, which ITE already knows how to do.

File: sympy/functions/elementary/piecewise.py [hunk 1]
```diff
@@ -546,6 +563,17 @@ def piecewise_fold(expr):
         for e, c in expr.args:
             if not isinstance(e, Piecewise):
                 e = piecewise_fold(e)
+            if c.has(Piecewise):
+                c = piecewise_fold(c)
+                if isinstance(c, Piecewise):
+                    # Piecewise((c1, k1), (c2, k2)) ->
```
Comment (smichr) [inline sympy/functions/elementary/piecewise.py]: ping @jksuom , can you give these few lines a once over, please? I am rewriting a Piecewise condition as an Or or Ands, e.g. `Piecewise((x < 1, y < 0), (x > 2, y < 1))` -> `(x < 1) & (y < 0) | (x > 2) & (y < 1) & ~(y < 0)`.

Comment (smichr) [reply sympy/functions/elementary/piecewise.py]: Perhaps this should become a method of Piecewise so one could write a Piecewise of conditions and then convert it into a Boolean. Perhaps `Piecewise.as_boolean()` or `to_nnf` as for `ITE`?

Comment (smichr) [reply sympy/functions/elementary/piecewise.py]: But I think this is the `dnf` form, so the method would be `to_dnf`.

**Comment by smichr** [comment]
It would be good to have someone take a look at the `to_dnf` method.

**Comment by smichr** [comment]
@asmeurer , what do you think about the `to_dnf` method? Should I call it `as_condition` instead? or something else? Or perhaps I could add an ITE rewrite method -- ITE has a PIecewise rewrite method -- and then use the `to_dnf` method from there.

**Comment by smichr** [comment]
If #13255 is added, rewriting a Piecewise with no default condition could be done by making the default condition be nan.

**Comment by smichr** [comment]
- [ ] Find out why this is failing:

```
 File "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/sympy/logic/tests/test_boolalg.py", line 800, in test_simplify_Boolean

assert simplify((x < oo) & (S(2) < x)) == (x > 2)

```

**Comment by smichr** [comment]
@asmeurer , if you can look at the OP to see if this is all in line with what you expect, that would be helpful.

File: sympy/logic/boolalg.py [hunk 2]
```diff
@@ -284,9 +363,71 @@ class BooleanFunction(Application, Boolean):
     It is used as base class for And, Or, Not, etc.
     """
     is_Boolean = True
+    def __lt__(self, other):
+        from sympy.logic.boolalg import Boolean
+        if isinstance(self, Boolean):
```
Comment (asmeurer) [inline sympy/logic/boolalg.py]: This will always be True. BooleanFunction is a subclass of Boolean.

Comment (smichr) [reply sympy/logic/boolalg.py]: Corrected and tested.

File: sympy/calculus/tests/test_util.py [hunk 3]
```diff
@@ -12,16 +12,30 @@
...
+        ) == Interval(-oo, oo)
+    assert function_range(1/(x**2), x, Interval(-1, 1)
+        ) == Interval(1, oo)
+    assert function_range(exp(x), x, Interval(-1, 1)
+        ) == Interval(exp(-1), exp(1))
+    assert function_range(log(x) - x, x, S.Reals
+        ) == Interval(-oo, -1)
+    assert function_range(sqrt(3*x - 1), x, Interval(0, 2)
+        ) == Interval(0, sqrt(5))
+    assert function_range(x*(x - 1) - (x**2 - x), x, S.Reals
```
Comment (smichr) [inline sympy/calculus/tests/test_util.py]: added this and the next one -- the others are just line wrapped

File: sympy/core/basic.py [hunk 4]
```diff
@@ -13,6 +13,22 @@
 from inspect import getmro
 
 
+def as_Basic(expr):
+    """Return expr as a Basic instance using strict sympify
+    or raise a TypeError"""
+    from sympy.utilities.misc import func_name
+    if not isinstance(expr, Basic):
```
Comment (smichr) [inline sympy/core/basic.py]: make this a simpler wrapper

File: sympy/core/relational.py [hunk 5]
```diff
@@ -230,32 +230,14 @@ def __nonzero__(self):
...
-        >>> x = Symbol('x', real=True)
-        >>> (x > 0).as_set()
-        Interval.open(0, oo)
-        >>> Eq(x, 0).as_set()
-        {0}
-
-        """
+    def _eval_as_set(self):
+        # self is univariate and periodicity(self, x) in (0, None)
+        from sympy.solvers.solveset import solveset
```
Comment (smichr) [inline sympy/core/relational.py]: remove this import

**Comment by smichr** [comment]
Add a test which, when it fair, will indicate that the 2.7 code can be removed.

**Pull request merged**